execution_engine/src/storage/trie/mod.rs

```diff
@@ -346,6 +350,96 @@ impl ::std::fmt::Debug for PointerBlock {
     }
 }
 
+/// Represents a Merkle Tree or a chunk of data with attached proof.
+/// Chunk with attached proof is used when the requested
+/// trie is larger than [ChunkWithProof::CHUNK_SIZE_BYTES].
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
+pub enum TrieOrChunkedData {
+    /// Represents a Merkle Trie.
+    Trie(Box<Trie<Key, StoredValue>>),
+    /// Represents a chunk of data with attached proof.
+    ChunkWithProof(ChunkWithProof),
+}
+
+impl Display for TrieOrChunkedData {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        match self {
+            TrieOrChunkedData::Trie(trie) => {
+                write!(f, "Trie({})", trie)
+            }
+            TrieOrChunkedData::ChunkWithProof(chunk) => {
+                write!(
+                    f,
+                    "ChunkWithProof {{ index = {}, hash = {} }}",
+                    chunk.proof().index(),
+                    chunk.proof().root_hash()
+                )
+            }
+        }
+    }
+}
+
+impl TrieOrChunkedData {
+    fn tag(&self) -> u8 {
+        match self {
+            TrieOrChunkedData::Trie(_) => 0,
+            TrieOrChunkedData::ChunkWithProof(_) => 1,
+        }
+    }
+}
+
+impl ToBytes for TrieOrChunkedData {
+    fn to_bytes(&self) -> Result<Vec<u8>, bytesrepr::Error> {
+        let mut ret = bytesrepr::allocate_buffer(self)?;
+        ret.push(self.tag());
+
+        match self {
+            TrieOrChunkedData::Trie(trie) => {
+                ret.append(&mut trie.to_bytes()?);
+            }
+            TrieOrChunkedData::ChunkWithProof(chunk) => {
+                ret.append(&mut chunk.to_bytes()?);
+            }
+        }
+        Ok(ret)
+    }
+
+    fn serialized_length(&self) -> usize {
+        U8_SERIALIZED_LENGTH
+            + match self {
+                TrieOrChunkedData::Trie(trie) => trie.serialized_length(),
+                TrieOrChunkedData::ChunkWithProof(chunk) => chunk.serialized_length(),
+            }
+    }
+}
+
+impl FromBytes for TrieOrChunkedData {
+    fn from_bytes(bytes: &[u8]) -> Result<(Self, &[u8]), bytesrepr::Error> {
+        let (tag, rem) = u8::from_bytes(bytes)?;
+        match tag {
+            0 => {
+                let (trie, rem) = Trie::<Key, StoredValue>::from_bytes(rem)?;
+                Ok((TrieOrChunkedData::Trie(Box::new(trie)), rem))
+            }
+            1 => {
+                let (chunk, rem) = ChunkWithProof::from_bytes(rem)?;
+                Ok((TrieOrChunkedData::ChunkWithProof(chunk), rem))
+            }
+            _ => Err(bytesrepr::Error::Formatting),
+        }
+    }
+}
+
+/// Represents the ID of a `TrieOrChunkedData` - containing the index and the root hash.
+#[derive(DataSize, Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq, Hash)]
+pub struct TrieOrChunkedDataId(pub u64, pub Digest);
```

Comment:
Is the second `Digest` element a hash of all chunks of that trie element? If that's true, can you include that explanation in the comment of `TrieOrChunkedDataId`?